### PR TITLE
Add "docking area signaller", a device for creating new white ship docks

### PR DESCRIPTION
**Pull request by warriorstar-orion**
## What Does This PR Do

This PR creates an object, the "docking area signaller", which allows players to create new docking ports for the white ship.

It only adds the implementation; it does not add it anywhere in the game. Once the explorer redux makes it in, and when/if the whiteship remap makes it in, it can be added, but it's not a good idea to add it now because the current whiteship has a wide airlock that will continuously move if the new docks are oriented north/south due to [ParadiseSS13/Paradise#24610](https://github.com/ParadiseSS13/Paradise/issues/24610).

I imagine once it does get added it'll show up on the whiteship itself or something.

In the future we may want to make it possible for players to create a new whiteship shuttle _console_ but for now the only whiteship consoles will remain on the ship itself, and at Centcomm.

When activated, it asks for a directional orientation, creates the port, checks the turfs returned by the port for suitability, and, if successful, resets all whiteship shuttle console destinations and forces them to recalculate the available destinations.

There are a ton of restrictions on this and I'm probably missing some:

1. It can only be used three times.
2. It can only be used once per sector.
3. It cannot be activated on the station z-level, Centcomm z-level, or Lavaland z-level.
4. All attempts to use it are adminlogged.
5. No objects or non-space turfs can be in the selected area.
6. It cannot be within the transition border range (7 tiles from any edge of the screen).

## Why It's Good For The Game

Giving explorers the ability to move the white ship to new specific areas is useful, and makes the ship more useful as a base of operations.

## Images of changes
Ignore the marker beacons.

https://github.com/ParadiseSS13/Paradise/assets/59303604/bfc868b8-a779-49dd-bb13-90ed42fecde5

## Testing

Spawned in, created ports, transited ship, attempted to spawn ports on station z-level, centcomm, and lavaland.

## Changelog

NPFC. Whatever PR actually adds the item in-game should have an appropriate changelog for it.

### CODE REVIEW

File: code/modules/shuttle/dock_generator.dm [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+		return
+
+	if(is_admin_level(user.z))
+		log_admin("[key_name(user)] attempted to create a whiteship dock on Centcomm level at [COORD(user)].")
+		to_chat(user, "A new docking area cannot be placed in this sector!")
+		return
+
+	if(is_mining_level(user.z))
+		log_admin("[key_name(user)] attempted to create a whiteship dock on Lavaland at [COORD(user)].")
+		to_chat(user, "A new docking area cannot be placed here!")
```
Comment (1080pCat) [inline code/modules/shuttle/dock_generator.dm:28]:
```suggestion
		to_chat(user, "New docking areas cannot be designated planet side!")
```

File: code/modules/shuttle/dock_generator.dm [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+/obj/item/whiteship_port_generator/attack_self(mob/living/user)
+	if(is_station_level(user.z))
+		log_admin("[key_name(user)] attempted to create a whiteship dock in the station's sector at [COORD(user)].")
+		to_chat(user, "New docking areas cannot be placed in the station's sector!")
+		return
+
+	if(is_admin_level(user.z))
+		log_admin("[key_name(user)] attempted to create a whiteship dock on Centcomm level at [COORD(user)].")
+		to_chat(user, "A new docking area cannot be placed in this sector!")
```
Comment (1080pCat) [inline code/modules/shuttle/dock_generator.dm:23]:
```suggestion
		to_chat(user, "New docking areas cannot be designated in this sector!")
```

File: code/modules/shuttle/dock_generator.dm [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+/obj/item/whiteship_port_generator/examine(mob/user)
+	. = ..()
+	var/plural = length(placed_docks) > 1
+	. += "There [plural ? "are" : "is"] [max_docks - length(placed_docks)] use[plural ? "s" : ""] left."
+
+/obj/item/whiteship_port_generator/attack_self(mob/living/user)
+	if(is_station_level(user.z))
+		log_admin("[key_name(user)] attempted to create a whiteship dock in the station's sector at [COORD(user)].")
+		to_chat(user, "New docking areas cannot be placed in the station's sector!")
```
Comment (1080pCat) [inline code/modules/shuttle/dock_generator.dm:18]:
```suggestion
		to_chat(user, "New docking areas cannot be designated within the station's sector!")
```

File: code/modules/shuttle/dock_generator.dm [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+	var/max_docks = 3
+
+/obj/item/whiteship_port_generator/examine(mob/user)
+	. = ..()
+	var/plural = length(placed_docks) > 1
+	. += "There [plural ? "are" : "is"] [max_docks - length(placed_docks)] use[plural ? "s" : ""] left."
+
+/obj/item/whiteship_port_generator/attack_self(mob/living/user)
+	if(is_station_level(user.z))
+		log_admin("[key_name(user)] attempted to create a whiteship dock in the station's sector at [COORD(user)].")
```
Comment (1080pCat) [inline code/modules/shuttle/dock_generator.dm:17]: is logging stuff that isn't/shouldn't be™️possible really needed?

File: code/modules/shuttle/dock_generator.dm [hunk 5]
```diff
@@ -0,0 +1,120 @@
+/obj/item/whiteship_port_generator
+	name = "docking area signaller"
+	desc = "A signaling device for the NT expeditionary vessel, used to assign new docking areas."
```
Comment (1080pCat) [inline code/modules/shuttle/dock_generator.dm:3]:
```suggestion
	desc = "A signaling device for the NT expeditionary vessel, used to designate new docking areas."
```

**Review by 1080pCat — COMMENTED** [review]
designating sounds *\~cooler~*

**Review by Henri215 — COMMENTED** [review]
to_chats missing span class

File: code/modules/shuttle/dock_generator.dm (hunk 4 above)
Comment (warriorstar-orion) [reply code/modules/shuttle/dock_generator.dm:17]: Yeah, we still want to see if people try to do griefy shit, even if it doesn't succeed.

File: code/modules/shuttle/dock_generator.dm [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	icon = 'icons/obj/device.dmi'
+	icon_state = "gangtool-white"
+
+	var/list/placed_docks = list()
+	var/max_docks = 3
+
+/obj/item/whiteship_port_generator/examine(mob/user)
+	. = ..()
+	var/plural = length(placed_docks) > 1
+	. += "There [plural ? "are" : "is"] [max_docks - length(placed_docks)] use[plural ? "s" : ""] left."
```
Comment (Henri215) [inline code/modules/shuttle/dock_generator.dm:13]: image

var/plural should be checking for "max_docks - length(placed_docks)" and not only "length(placed_docks)" right?

Comment (warriorstar-orion) [reply code/modules/shuttle/dock_generator.dm]: Yup, fixed.

**Review by Henri215 — APPROVED** [review]
Lewcs good to me

**Review by Burzah — APPROVED** [review]

**Pull request merged**